knimin/lib/data_access.py

```diff
@@ -2759,6 +2767,155 @@ def delete_samples(self, ids):
+                    raise ValueError('Sample ID %s does not exist.'
+                                     % sample_id)
+                if self._study_sample_are_associated(study_id, sample_id):
+                    raise ValueError('Study ID %s and sample ID %s are already'
+                                     ' associated.' % (study_id, sample_id))
+                sql = """INSERT INTO pm.study_sample (study_id, sample_id)
+                         VALUES (%s, %s)"""
+                TRN.add(sql, [study_id, sample_id])
+
+    def dissociate_samples_from_study(self, study_id, sample_ids):
+        """Deletes the associations between one or more samples and one study
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+        sample_ids : list of str
+            IDs of the samples
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+            If the study and the sample are not associated
+        """
+        self._study_exists(study_id)
+        with TRN:
+            for sample_id in sample_ids:
+                if not self._sample_exists(sample_id):
+                    raise ValueError('Sample ID %s does not exist.'
+                                     % sample_id)
+                if not self._study_sample_are_associated(study_id, sample_id):
+                    raise ValueError('Study ID %s and sample ID %s are not '
+                                     'associated.' % (study_id, sample_id))
+                sql = """DELETE FROM pm.study_sample
+                         WHERE study_id = %s AND sample_id = %s"""
+                TRN.add(sql, [study_id, sample_id])
+
+    def get_samples_associated_with_study(self, study_id):
+        """Retrieves IDs of all samples associated with a study (if any)
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+
+        Returns
+        ------
+        list of str
+            Sorted list of sample IDs associated with the study
+        """
+        self._study_exists(study_id)
+        sql = """SELECT sample_id FROM pm.study_sample WHERE study_id = %s
+                 ORDER BY sample_id"""
+        res = self._con.execute_fetchall(sql, [study_id])
+        if res is not None:
+            return [x[0] for x in res]
+        else:
+            return []
+
+    def get_studies_associated_with_samples(self, sample_ids):
+        """Retrieves IDs of all samples associated with a study (if any)
+
+        Parameters
+        ----------
+        sample_id : list of str
+            IDs of the samples
+
+        Returns
+        ------
+        list of list of int
+            List of sorted list of study IDs associated with each sample in
+            the same order as sample IDs
+
+        Raises
+        ------
+        ValueError
+            If at least one sample ID does not exist
+        """
+        associations = []
```

Comment:
I think this entire function can be replaced with (probably a couple of tweeks with the `association` variable to match types:
```python
sql = """SELECT sample_id, array_agg(study_id ORDER BY study_id)
         FROM pm.study_sample
         WHERE sample_id IN (%s)"""
TRN.add(sql, [tuple(sample_ids])
associations = TRN.execute_fetchindex()
```

Any special reason for returning a list of list of int rather than a dict of `{sample_id: [study_ids]}`. I have the feeling that relying on sorting is fragile.